src/Servers/HttpSys/src/RequestProcessing/Response.cs

```diff
@@ -660,76 +613,65 @@ internal unsafe void SerializeTrailers(HttpApiTypes.HTTP_DATA_CHUNK[] dataChunks
-
-                    // Add Value
-                    var headerValue = headerValues[headerValueIndex] ?? string.Empty;
-                    bytes = HeaderEncoding.GetBytes(headerValue);
-                    unknownHeaders[unknownHeadersOffset].RawValueLength = (ushort)bytes.Length;
-                    gcHandle = GCHandle.Alloc(bytes, GCHandleType.Pinned);
-                    pinnedHeaders.Add(gcHandle);
-                    unknownHeaders[unknownHeadersOffset].pRawValue = (byte*)gcHandle.AddrOfPinnedObject();
-                    unknownHeadersOffset++;
-                }
+            for (int headerValueIndex = 0; headerValueIndex < headerValues.Count; headerValueIndex++)
+            {
+                // Add Name
+                var bytes = allocator.GetHeaderEncodedBytes(headerName, out int bytesLength);
+                unknownHeaders[unknownHeadersOffset].NameLength = checked((ushort)bytesLength);
+                unknownHeaders[unknownHeadersOffset].pName = bytes;
+
+                // Add Value
+                var headerValue = headerValues[headerValueIndex] ?? string.Empty;
+                bytes = allocator.GetHeaderEncodedBytes(headerValue, out bytesLength);
+                unknownHeaders[unknownHeadersOffset].RawValueLength = checked((ushort)bytesLength);
+                unknownHeaders[unknownHeadersOffset].pRawValue = bytes;
+                unknownHeadersOffset++;
             }
-
-            Debug.Assert(unknownHeadersOffset == trailerCount);
-        }
-        catch
-        {
-            FreePinnedHeaders(pinnedHeaders);
-            throw;
         }
 
-        // Success, keep the pins.
-        pins.AddRange(pinnedHeaders);
+        Debug.Assert(unknownHeadersOffset == trailerCount);
     }
 
     // Subset of ComputeHeaders
     internal void SendOpaqueUpgrade()
     {
         _boundaryType = BoundaryType.Close;
 
-        // TODO: Send headers async?
-        ulong errorCode = SendHeaders(null, null,
-            HttpApiTypes.HTTP_FLAGS.HTTP_SEND_RESPONSE_FLAG_OPAQUE |
-            HttpApiTypes.HTTP_FLAGS.HTTP_SEND_RESPONSE_FLAG_MORE_DATA |
-            HttpApiTypes.HTTP_FLAGS.HTTP_SEND_RESPONSE_FLAG_BUFFER_DATA,
-            true);
-
-        if (errorCode != ErrorCodes.ERROR_SUCCESS)
+        UnmanagedBufferAllocator allocator = new();
+        try
+        {
+            // TODO: Send headers async?
+            ulong errorCode = SendHeaders(ref allocator, null, null,
+                HttpApiTypes.HTTP_FLAGS.HTTP_SEND_RESPONSE_FLAG_OPAQUE |
+                HttpApiTypes.HTTP_FLAGS.HTTP_SEND_RESPONSE_FLAG_MORE_DATA |
+                HttpApiTypes.HTTP_FLAGS.HTTP_SEND_RESPONSE_FLAG_BUFFER_DATA,
+                true);
+
+            if (errorCode != ErrorCodes.ERROR_SUCCESS)
+            {
+                throw new HttpSysException((int)errorCode);
+            }
+        }
+        finally
         {
-            throw new HttpSysException((int)errorCode);
+            allocator.Dispose();
```

Comment:
Any reason for an explicit dispose call here?

Reply:
The `using` pattern isn't possible when taking a `ref` to the variable in the pattern. Note there is one place where the `using` was possible in this review.

Reply:
Add a comment?